diff -Naur a/ft245r.c b/ft245r.c
--- a/ft245r.c	2014-09-16 03:01:09.000000000 +0800
+++ b/ft245r.c	2020-02-22 21:00:23.381036600 +0800
@@ -653,14 +653,22 @@
 
 
 static void ft245r_close(PROGRAMMER * pgm) {
+    int retry_times = 0;
     if (handle) {
         // I think the switch to BB mode and back flushes the buffer.
         ftdi_set_bitmode(handle, 0, BITMODE_SYNCBB); // set Synchronous BitBang, all in puts
         ftdi_set_bitmode(handle, 0, BITMODE_RESET); // disable Synchronous BitBang
         ftdi_usb_close(handle);
-        ftdi_deinit (handle);
-        pthread_cancel(readerthread);
+        while(pthread_cancel(readerthread) && retry_times < 100) {
+            retry_times++;
+            usleep(100);
+        }
+        if (retry_times >= 100) {
+            avrdude_message(MSG_INFO, "Too many retry to close reader thread\n");
+        }
+
         pthread_join(readerthread, NULL);
+        ftdi_deinit (handle);
         free(handle);
         handle = NULL;
     }
