--- avrdude.conf.in.orig	2016-12-16 13:08:00.210983367 +0100
+++ avrdude.conf.in	2016-12-16 13:10:44.691541657 +0100
@@ -6548,6 +6548,13 @@
                           "x x x x  x x x x   i i i i  i i i i";
       ;
 
+    # Required for Arduino IDE
+    # see: https://github.com/arduino/Arduino/issues/2075
+    #      https://github.com/arduino/Arduino/issues/2075#issuecomment-238031689
+    memory "efuse"
+        size            = 0;
+      ;
+
     memory "lock"
         size            = 1;
         min_write_delay = 2000;
